routing/payment_lifecycle_test.go

```diff
@@ -491,3 +662,943 @@ func TestDecideNextStep(t *testing.T) {
+	t.Parallel()
+
+	// Create a test paymentLifecycle with the initial two calls mocked.
+	p, m := newTestPaymentLifecycle(t)
+
+	paymentAmt := 10_000
+	preimage := lntypes.Preimage{1}
+	attempt := makeSettledAttempt(t, paymentAmt, preimage)
+
+	// Mock shardTracker to return the payment hash.
+	m.shardTracker.On("GetHash",
+		attempt.AttemptID,
+	).Return(p.identifier, nil).Once()
+
+	// Mock the htlcswitch to return a the result chan.
+	resultChan := make(chan *htlcswitch.PaymentResult, 1)
+	m.payer.On("GetAttemptResult",
+		attempt.AttemptID, p.identifier, mock.Anything,
+	).Return(resultChan, nil).Once().Run(func(args mock.Arguments) {
+		// Send the preimage to the result chan.
+		resultChan <- &htlcswitch.PaymentResult{
+			Preimage: preimage,
+		}
+	})
+
+	// Once the result is received, `ReportPaymentSuccess` should be
+	// called.
+	m.missionControl.On("ReportPaymentSuccess",
+		attempt.AttemptID, &attempt.Route,
+	).Return(nil).Once()
+
+	// Now the settled htlc being returned from `SettleAttempt`.
+	m.control.On("SettleAttempt",
+		p.identifier, attempt.AttemptID, mock.Anything,
+	).Return(attempt, nil).Once()
+
+	// Mock the clock to return a current time.
+	m.clock.On("Now").Return(time.Now())
+
+	// Now call the method under test.
+	result, err := p.collectResult(attempt)
+	require.NoError(t, err, "expected no error")
+	require.Equal(t, preimage, result.attempt.Settle.Preimage,
+		"preimage mismatch")
+}
+
+// TestCollectResultAsyncSuccess checks a successful htlc settlement.
+func TestCollectResultAsyncSuccess(t *testing.T) {
+	t.Parallel()
+
+	// Create a test paymentLifecycle with the initial two calls mocked.
+	p, m := newTestPaymentLifecycle(t)
+
+	paymentAmt := 10_000
+	preimage := lntypes.Preimage{1}
+	attempt := makeSettledAttempt(t, paymentAmt, preimage)
+
+	// Mock shardTracker to return the payment hash.
+	m.shardTracker.On("GetHash",
+		attempt.AttemptID,
+	).Return(p.identifier, nil).Once()
+
+	// Mock the htlcswitch to return a the result chan.
+	resultChan := make(chan *htlcswitch.PaymentResult, 1)
+	m.payer.On("GetAttemptResult",
+		attempt.AttemptID, p.identifier, mock.Anything,
+	).Return(resultChan, nil).Once().Run(func(args mock.Arguments) {
+		// Send the preimage to the result chan.
+		resultChan <- &htlcswitch.PaymentResult{
+			Preimage: preimage,
+		}
+	})
+
+	// Once the result is received, `ReportPaymentSuccess` should be
+	// called.
+	m.missionControl.On("ReportPaymentSuccess",
+		attempt.AttemptID, &attempt.Route,
+	).Return(nil).Once()
+
+	// Now the settled htlc being returned from `SettleAttempt`.
+	m.control.On("SettleAttempt",
+		p.identifier, attempt.AttemptID, mock.Anything,
+	).Return(attempt, nil).Once()
+
+	// Mock the clock to return a current time.
+	m.clock.On("Now").Return(time.Now())
+
+	// Now call the method under test.
+	p.collectResultAsync(attempt)
+
+	// Assert the result is returned within 5 seconds.
+	var err error
+	require.Eventually(t, func() bool {
```

Comment:
yeah I think so. So it's call the inner func every 10ms, which leads me to think our `wait` package is better as it will block on the first goroutine, switched.